benchllm/cache.py

```diff
@@ -0,0 +1,87 @@
+import json
+from pathlib import Path
+from typing import Optional
+
+from benchllm.data_types import Evaluation, Prediction
+from benchllm.evaluator import Evaluator
+from benchllm.input_types import Json
+from benchllm.listener import EvaluatorListener
+
+
+class MemoryCache(Evaluator):
+    """Caches the results of the evaluator in memory"""
+
+    def __init__(self, evaluator: Evaluator):
+        super().__init__(workers=evaluator.workers)
+        self._data: dict = {}
+        self._evaluator = evaluator
+        self._num_cache_misses = 0
+        self._num_cache_hits = 0
+
+    def _key(self, answer1: Json, answer2: Json) -> str:
+        key1, key2 = json.dumps([answer1, answer2]), json.dumps([answer2, answer1])
+        return key1 if key1 < key2 else key2
+
+    def lookup(self, answer1: Json, answer2: Json) -> Optional[bool]:
+        return self._data.get(self._key(answer1, answer2), None)
+
+    def store(self, answer1: Json, answer2: Json, value: bool) -> None:
+        key = self._key(answer1, answer2)
+        self._data[key] = value
+
+    def evaluate_prediction(self, prediction: Prediction) -> Optional[Evaluator.Match]:
+        for expected in prediction.test.expected:
+            lookup = self.lookup(expected, prediction.output)
+            # None indicates that nothing was found in the cache
+            # while True and False are both valid cache values
+            if lookup is None:
+                continue
+            self._num_cache_hits += 1
+            if lookup:
+                return Evaluator.Match(prediction=prediction.output, expected=expected)
+            return None
```

Comment:
I am not sure I get the logic here. If we miss, we want to break for this loop, not return, right? Otherwise, we don't count the miss (line 44).

Reply:
```python
for expected in prediction.test.expected:
            lookup = self.lookup(expected, prediction.output)
            # None indicates that nothing was found in the cache
            # while True and False are both valid cache values
            if lookup is None:
                continue
            self._num_cache_hits += 1
            if lookup:
                return Evaluator.Match(prediction=prediction.output, expected=expected)
            return None
```

If lookup is None, then we don't have any entry for that prediction, expected tuple so we continue. 
If we have an entry, it might be a positive entry (they match) and then we can stop searching. 
If we have an entry, it might also be a negative entry (they do not match) and we can also stop searching [this could be debated tbh]

Reply:
Have a look now, your comment helped me find a corner case which I also wrote a test case for!

Reply:
I see, thank you for adding some documentation, it's much clearer now!